plugins/UM3NetworkPrinting/tests/TestSendMaterialJob.py

```diff
@@ -0,0 +1,197 @@
+
+        # We expect the error string to be retrieved and the device not to be called for any follow up.
+        self.assertEqual([call.attribute(0), call.errorString()], reply_mock.method_calls)
+        self.assertEqual(0, device_mock.createFormPart.call_count)
+
+    def test__onGetRemoteMaterials_withBadJsonAnswer(self, reply_mock, device_mock):
+        reply_mock.attribute.return_value = 200
+        reply_mock.readAll.return_value = QByteArray(b"Six sick hicks nick six slick bricks with picks and sticks.")
+        job = SendMaterialJob(device_mock)
+        job._onGetRemoteMaterials(reply_mock)
+
+        # We expect the reply to be called once to try to get the printers from the list (readAll()).
+        # Given that the parsing fails we do no expect the device to be called for any follow up.
+        self.assertEqual([call.attribute(0), call.readAll()], reply_mock.method_calls)
+        self.assertEqual(0, device_mock.createFormPart.call_count)
+
+    def test__onGetRemoteMaterials_withMissingGuidInRemoteMaterial(self, reply_mock, device_mock):
+        reply_mock.attribute.return_value = 200
+        remote_material_without_guid = self._REMOTE_MATERIAL_WHITE.copy()
+        del remote_material_without_guid["guid"]
+        reply_mock.readAll.return_value = QByteArray(json.dumps([remote_material_without_guid]).encode("ascii"))
+        job = SendMaterialJob(device_mock)
+        job._onGetRemoteMaterials(reply_mock)
+
+        # We expect the reply to be called once to try to get the printers from the list (readAll()).
+        # Given that parsing fails we do not expect the device to be called for any follow up.
+        self.assertEqual([call.attribute(0), call.readAll()], reply_mock.method_calls)
+        self.assertEqual(0, device_mock.createFormPart.call_count)
+
+    @patch("cura.Settings.CuraContainerRegistry")
+    @patch("cura.CuraApplication")
+    def test__onGetRemoteMaterials_withInvalidVersionInLocalMaterial(self, application_mock, container_registry_mock,
+                                                                     reply_mock, device_mock):
+        reply_mock.attribute.return_value = 200
+        reply_mock.readAll.return_value = QByteArray(json.dumps([self._REMOTE_MATERIAL_WHITE]).encode("ascii"))
+
+        localMaterialWhiteWithInvalidVersion = self._LOCAL_MATERIAL_WHITE.copy()
+        localMaterialWhiteWithInvalidVersion["version"] = "one"
+        container_registry_mock.findContainersMetadata.return_value = [localMaterialWhiteWithInvalidVersion]
+
+        application_mock.getContainerRegistry.return_value = container_registry_mock
+
+        with mock.patch.object(CuraApplication, "getInstance", new = lambda: application_mock):
+            job = SendMaterialJob(device_mock)
+            job._onGetRemoteMaterials(reply_mock)
+
+        self.assertEqual([call.attribute(0), call.readAll()], reply_mock.method_calls)
+        self.assertEqual([call.getContainerRegistry()], application_mock.method_calls)
```

Comment:
You're testing an implementation detail with these assertions. It doesn't matter that the `Application.getContainerRegistry()` method is called; it's not a requirement of the `SendMaterialJob`. Instead test the results, so that if we choose to change the implementation, the test won't break.

This goes for most of the assertions against `method_calls` in these tests. The ones that test against `createFormPart` and such are the only ones that actually test the public behaviour.

Reply:
I agree. But that is the only way we can verify the methods. This is the best we can do now.

I think this a result of retrofitting tests on a class that not developed test-first. No consideration is given to how to check for the correct working of the code when the code was written.

Reply:
If it's the only way to verify the methods, then don't. The alternative is far worse; Checking behavior that doesn't matter if it's done in that way. 
It's about testing the result, not about testing the way it did something.

Reply:
Ok, I removed the asserts on the internals and only look at the output.